helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java

```diff
@@ -0,0 +1,181 @@
+          for (String p : _pattern) {
+            if (target.toLowerCase().contains(p.toLowerCase())) {
+              return true;
+            }
+          }
+          return false;
+        };
+        return pred;
+      }
+
+      List<Predicate<String>> predicateList = new ArrayList<>();
+      for (ThreadCategory threadCategory : ThreadCategory.values()) {
+        if (threadCategory == ThreadCategory.Other) {
+          continue;
+        }
+        predicateList.add(threadCategory.getMatchPred());
+      }
+      Predicate<String> pred = target -> {
+        for (Predicate<String> p : predicateList) {
+          if (p.test(target)) {
+            return false;
+          }
+        }
+        return true;
+      };
+
+      return pred;
+    }
+
+    public int getWarningLimit() {
+      return _warningLimit;
+    }
+
+    public int getLimit() {
+      return _limit;
+    }
+
+    private ThreadCategory(String description, int warningLimit, int limit, String[] patterns) {
+      _description = description;
+      _pattern = Arrays.asList(patterns);
+      _warningLimit = warningLimit;
+      _limit = limit;
+    }
+  }
+
+  public static boolean afterClassCheck(String classname) {
+    ZkTestBase.reportPhysicalMemory();
+    // step 1: get all active threads
+    List<Thread> threads = getAllThreads();
+    System.out.println(classname + " has active threads cnt:" + threads.size());
+
+    // step 2: categorize threads
+    Map<String, List<Thread>> threadByName = null;
+    Map<ThreadCategory, Integer> threadByCnt = new HashMap<>();
+    Map<ThreadCategory, Set<Thread>> threadByCat = new HashMap<>();
+    try {
+      threadByName = threads.
+          stream().
+          filter(p -> p.getThreadGroup() != null && p.getThreadGroup().getName() != null
+              &&  ! "system".equals(p.getThreadGroup().getName())).
+          collect(Collectors.groupingBy(p -> p.getName()));
+    } catch (Exception e) {
+      System.out.println("filtering thread failure with exception:" + e.getStackTrace());
+    }
+
+    threadByName.entrySet().stream().forEach(entry -> {
+      String key = entry.getKey(); // thread name
+      Arrays.asList(ThreadCategory.values()).stream().forEach(category -> {
+        if (category.getMatchPred().test(key)) {
+          Integer count = threadByCnt.containsKey(category) ? threadByCnt.get(category) : 0;
+          threadByCnt.put(category, count + entry.getValue().size());
+          Set<Thread> thisSet = threadByCat.getOrDefault(category, new HashSet<>());
+          thisSet.addAll(entry.getValue());
+          threadByCat.put(category, thisSet);
+        }
+      });
+    });
+
+    // step 3: enforce checking policy
+    boolean checkStatus = true;
+    for (ThreadCategory threadCategory : ThreadCategory.values()) {
+      int limit = threadCategory.getLimit();
+      int warningLimit = threadCategory.getWarningLimit();
+
+      Integer catThreadCnt = threadByCnt.get(threadCategory);
```

Comment:
nit: can you use more character for "category", catThreadCnt is not quit self explanatory.

Reply:
changed